### ISSUE TITLE
[Feature request] Support indexing nested fields

### ISSUE THREAD

**Issue by goleary** [typesense/typesense#227]
I'd like to set up my collection like this

```js
const schema = {
  name: 'links',
  num_documents: 0,
  fields: [
    {
      name: 'company.name',
      type: 'string',
      facet: false,
    },
    {
      name: 'employees.num',
      type: 'int32',
      facet: false,
    },
    {
      name: 'country.code',
      type: 'string',
      facet: true,
    },
  ],
  default_sorting_field: 'employees.num',
}
```

And be able to index objects like this
```js
const indexedObj = {
  company: {
    name: 'nike';
  },
  employees: {
  	num: 1200;
  },
  country: {
    code: 'us'
  }
}
```

The reason that I want to do this is that I want to preserve my data model in objects that I store in TypeSense so I don't have do any transformation when I get them back from Typesense.

Without being able to index these fields I have to duplicate data from the nested object into the top level of objects.

**Comment by joshmossas** [comment]
I would also very much like this feature.  We have 3.5 million records in Algolia right now and it's costing a pretty penny, but our data model has a lot of nested fields. This feature would make migration a lot simpler for sure.

We also have facets that are within those nested fields. Including facets that are from fields within arrays of objects.

For example:
```ts
interface SearchObject {
  objectID: string;
  type: {
    id: string; // this is a facet
    name: string;
    description: string;
  };
  specialty: {
    id: string; // this is a facet
    name: string;
    description: string;
    medicareTaxonomyName: string; // this is a facet
    medicareTaxonomyCode: string; // this is a facet
  };
  locations: [
    {
       id: string; // this is a facet
       name: string;
       cityState: string; // this is a facet
       address: {
         line1: string;
         line2: string;
         city: string;
         region: string; // this is a facet
       };
    }
  ]
}
```

**Comment by kylewill** [comment]
Possible to review this feature request? I'm running into the same challenge migrating from Algolia

**Comment by goleary** [comment]
According to @jasonbosco the team _is_ working on it although it likely won't make it into the next release (v0.21.0).

**Comment by jasonbosco** [comment]
Yeah, we thought we could get this out in v0.21.0, but supporting nested arrays of objects turned into a rabbit hole. So we've deferred this for the next release. 

Until then, the work around would be to flatten keys before indexing them in Typesense.

**Comment by HuynhVuong87** [comment]
can v0.22.0 index nested fields?
i am waiting for this, hope Typesense team will implement this soon

**Comment by kishorenc** [comment]
Not yet, unfortunately it's a significant and time consuming change. Given the intensive and long 0.22 release cycle, we want to now focus on shipping some important lower hanging fruits for the 0.23 release. We will probably tackle it in 0.24 release. But I really want to get to this soon.

**Comment by renehauck** [comment]
I currently have a similar problem. :-/ Wouldn't it be easier to go the SQL way and implement the join approach in Typesense instead?

**Comment by NotNoah** [comment]
Just got here from the docs, for anyone looking for the current solution, it's to flatten the nested fields - https://typesense.org/docs/0.22.1/api/collections.html#indexing-nested-fields

Would love to see this feature soon, it would be very useful!

**Comment by Vichet97** [comment]
While we are at it. Please also check if there's any possibility to query the unknown nested field.

`example: {  [unknown id]: "string"  } or even { "id": {  [unknown id]: "string"  } }`

Also, I would like to know if it's possible to search within the `string[]`

> Not yet, unfortunately it's a significant and time consuming change. Given the intensive and long 0.22 release cycle, we want to now focus on shipping some important lower hanging fruits for the 0.23 release. We will probably tackle it in 0.24 release. But I really want to get to this soon.

**Comment by iamclaytonray** [comment]
Would like to check on the status of this. Only thing holding me back from migrating several projects from Algolia to Typesense

**Comment by delzennejc** [comment]
This seems so basic and it makes me think twice about adopting typsense.

**Comment by PumpkinSeed** [comment]
Thanks for the amazing product. Is there any chance it will be available in the next few weeks? We have a client looking to adopt it with over hundreds of millions embedded documents.

**Comment by kishorenc** [comment]
We've started working on it and should be able to share an early preview build in a few weeks.

**Comment by marcospassos** [comment]
That's the most wanted feature in Typesense for us.

**Comment by Dararii** [comment]
> We've started working on it and should be able to share an early preview build in a few weeks.

How long to develop this feature? We want to migrate to Typesense but we need that feature before we proceed with the migration.

**Comment by kishorenc** [comment]
@Dararii I've been continuing to work on this, I'm about 75% done -- I think it will take a couple more weeks. I appreciate everyone's patience!

**Comment by Dararii** [comment]
> @Dararii I've been continuing to work on this, I'm about 75% done -- I think it will take a couple more weeks. I appreciate everyone's patience!

Thanks for the information. Will it support array of object too?

**Comment by kishorenc** [comment]
Yes! Array of objects will be supported.

**Comment by joeypunzel** [comment]
Very much looking forward to this feature - going to be a gamechanger for Typesense :) .

@jasonbosco @kishorenc just checking in -- is this feature still slated to be released with v0.24 this week or next? Any insight would be appreciated

**Comment by BreiteSeite** [comment]
> @Dararii I've been continuing to work on this, I'm about 75% done -- I think it will take a couple more weeks. I appreciate everyone's patience!

Hi @Dararii thank you for your work on this. Indeed, it's a nice feature.

One question i have though - does this have any implications on (query) performance and/or memory usage? Any benchmark?

If there is a (relatively) big impact on performance, i personally might consider still normalizing while indexing. :)

Thanks

**Comment by kishorenc** [comment]
Good news: I finally have a preview build that contains support for indexing and searching on nested fields!

Docker image: `typesense/typesense:0.24.0.nest1`

**Usage:**

 For now, support for nested fields **must be enabled** at a collection level via the `enable_nested_fields` schema property:

```
curl -k "http://localhost:8108/collections" -X POST -H "Content-Type: application/json" \
      -H "X-TYPESENSE-API-KEY: ${TYPESENSE_API_KEY}" -d '{
        "name": "docs", 
        "enable_nested_fields": true,
        "fields": [
          {"name": ".*", "type": "auto"}
        ]
      }'
```

Given the sheer amount of changes that this feature has involved, we want to do more testing under real world datasets before enabling it by default.

The schema can also explicitly index specific object fields or object arrays, e.g.:

```
curl -k "http://localhost:8108/collections" -X POST -H "Content-Type: application/json" \
      -H "X-TYPESENSE-API-KEY: ${TYPESENSE_API_KEY}" -d '{
        "name": "docs", 
        "enable_nested_fields": true,
        "fields": [
          {"name": "person", "type": "object"},
          {"name": "details", "type": "object[]"},
        ]
      }'
```

When you search on an object field name, all sub-fields will be automatically searched.

**Highlighting response structure**

Given the challenges of highlighting a nested object, we had to introduce a new `highlight` property in the response which handles highlighting. The older `highlights` array will continue to be provided, but it will not contain highlights of nested fields.

The nested highlight structure looks like this:

```
"higlight": {
      "employees":{
        "tags":[
          "senior plumber",
          "<mark>electrician</mark>"
        ]
      },
      "locations":[
        {
          "address":{
            "street":"One Bowerman Drive"
          }
        },
        {
          "address":{
            "street":"175 <mark>Commerce</mark> Valley"
          }
        }
      ]
    }
```

**Caveats / what's missing**

We've not implemented support for infix search and schema updates yet.

Please give it a spin and let us know how it works. We would like to polish and bring this feature to the finish line over the next few weeks with all your feedback!

**Comment by reinoldus** [comment]
Yess! Finally! Thank you!

I am toying with it already. For my use case I'd like to filter for two properties within an object that is part of an array of objects

**For example:**

Document:
```
{
    "name": "Chili con carne",
   "ingredients": [{
       "name": "Ground Beef",
       "concentration": 30
}]
}
```

I'd like to search for this meal like this:

```
search_parameters = {
    'q':              '*',
    'query_by':       'ingredients',
    'filter_by':      'ingredients.name: Ground Beef && ingredients.concentration:>= 20',
    'include_fields': 'ingredients'
}
```


So I'd like to filter for an ingredient that has a concentration of "Ground beef" of at least 20%.

How would I do that? Or isn't it possible right now, if so will it be possible?

The query above seems to look for all concentration estimates within the ingredients array and not just the one within the "ground beef" object.

Let me know if this isn't the right place for feedback.

**Comment by kishorenc** [comment]
@reinoldus This requires a new filter by syntax where we can scope a bunch of filter expressions into a given object's sub fields. There's some ongoing work on supporting cross field ORs and complex filter expressions. We should be able to get to this post that. But certainly possible!

**Comment by eminos** [comment]
Are there currently any direct known drawbacks compared to flattening?

**Comment by MantraMedia** [comment]
This is great news.

Maybe I am missing the obvious but is it also possible to have an explicit schema for nested fields and not just `auto/object`?

I am asking as I tried 

`{
      "name": ".*_facet",
      "type": "auto",
      "facet": true
    }`

to have facet fields in the nested object but they are not getting recognized as such.

**Comment by kishorenc** [comment]
@eminos 

> Are there currently any direct known drawbacks compared to flattening?

Under the hood, if you don't flatten and index an object as a whole, Typesense does the flattening on your behalf. There's a small overhead associated with this operation, and likewise for highlighting on a nested object (compared to a flat field), but when you think end to end, it's only a question of "where" this additional processing is done: with manual flattening of fields, that processions is done outside Typesense. People find it easier to deal with nested fields directly, in some cases like nested array of objects.

@MantraMedia 

> is it also possible to have an explicit schema for nested fields and not just auto/object?

You can index and search on specific sub-fields of an object, if you don't wish to index a whole object. For e.g. to index only the `name` field of a `company` object, you can have:

```
 {"name": "company.name", "type": "string", "optional": false }
```

> { "name": ".*_facet", "type": "auto", "facet": true }
> to have facet fields in the nested object but they are not getting recognized as such.

Can you post a sample document with the fields you are looking to target with this definition?

**Comment by e-lobo** [comment]
> `{"name": "company.name", "type": "string", "optional": false }`

what if we have an **_array of objects_** and just want to index certain fields? can we use the wildcard operator as well?

```
{
"product_name": "Pepsi", 
"variants": [ 
{"product_name": "pepsi-small", "product_name_ar": "بيبسي صغير",  description: "Soft drink." , ...extra_fields}  
{"product_name": "pepsi-large", "product_name_ar": "بيبسي كبير",  description: "Soft drink large.", ...extra_fields}  
]
```

In the above i want to index like so

`{"name": "variants.product_name.*", "type": "auto", "optional": True}`

&

`{"name": "variants.description", "type": "string", "optional": True}`

does this work? im not able to see results when i gave this a try

**Comment by kishorenc** [comment]
@e-lobo That's a good catch: we don't currently support referring to nested fields of the `variants` array by using a `variants.product_name.*` notation.

**Comment by WahidinAji** [comment]
> are an early preview b

is it only on Docker image?
because I didn't see on [these tags](https://github.com/typesense/typesense/tags)

**Comment by t-morgan** [comment]
@WahidinAji I am able to download the releases from https://dl.typesense.org/releases/ using the release candidate versions that are also available as Docker images.

**Comment by WahidinAji** [comment]
> @WahidinAji I am able to download the releases from https://dl.typesense.org/releases/ using the release candidate versions that are also available as Docker images.
> 
> e.g.
> 
> ```
> curl -O https://dl.typesense.org/releases/0.24.0.rcn13/typesense-server-0.24.0.rcn13-amd64.deb
> ```

Thanks, @t-morgan .
now, I can fix [this issue](https://github.com/typesense/typesense-go/issues/113) with the `object[]` type and of course with the 0.24 version

**Comment by kishorenc** [comment]
@WahidinAji 

I've just published `0.24.0.rcn14` which contains a few more fixes for edge cases reported, including the ability to alter schema for nested fields.

**Comment by mahmudtopu3** [comment]
when it will be released as a production version?

**Comment by wennals** [comment]
I am having an issues with the response of the snippets. I am no longer getting the markup in the Typesense response.
Here is a sample query and response:


QUERY: {
  q: 'book',
  query_by: 'cccedict.pinyin, cccedict.english, unihan.english',
  per_page: 250,
  use_cache: true,
  num_typos: 2,
  prefix: 'false, true, true',
  prioritize_token_position: true,
  highlight_fields: 'cccedict.english, unihan.english',
  page: 1
}

RESPONSE
{
...
  "highlight": 
                "full": {},
                "meta": {},
                "snippet": {
                    "cccedict.english": [
                        "impression of a book",
                        "opinion expressed in a book review"
                    ]
                }
            },
            "highlights": [],
            "text_match": 578730123365699700,
            "text_match_info": {
                "best_field_score": "1108091339002",
                "best_field_weight": 14,
                "fields_matched": 1,
                "score": "578730123365699697",
                "tokens_matched": 1
            }
        },

**Comment by kishorenc** [comment]
@wennals would it be possible for you to share a small dataset where this issue is reproduceable? You can email us at contact@typesense.org

**Comment by LuukOV** [comment]
It won't seem to index my nested object. 
`{
      "name": "propname",
      "type": "object[]",
      "facet": false,
      "optional": true,
      "index": true,
      "sort": false,
      "infix": false,
      "locale": "",
      "nested": true,
      "nested_array": 2
      }`
      I'm using version **0.24.0.rcn25**.
     I've tried putting it to auto, or using the latest version. but it will just never return the field within my results.
     example of the data I would have:
     `
     prop1: true,
     prop2: true,
     propname: [
     { type: "one", value: "two" },
      { type: "two", value: "two" }
     ]
     `

**Comment by AlexandreLage** [comment]
@kishorenc thanks for your hard work.

Is `enable_nested_fields` available on Typesense cloud yet?

**Comment by reinoldus** [comment]
When I reboot typesense/typesense:0.24.0.rcn28 I have to re-index my nested collections to make the filtering work again.

I have the issue with two collections, so I don't think it is data dependent. This is a single node local docker dev server.

Here is the boot log: 
```
I20221107 07:33:44.002331     1 typesense_server_utils.cpp:354] Starting Typesense 0.24.0.rcn28
I20221107 07:33:44.002380     1 typesense_server_utils.cpp:357] Typesense is using jemalloc.
I20221107 07:33:44.002645     1 typesense_server_utils.cpp:406] Thread pool size: 64
I20221107 07:33:44.024477     1 store.h:63] Initializing DB by opening state dir: /data/db
I20221107 07:33:44.359279     1 store.h:63] Initializing DB by opening state dir: /data/meta
I20221107 07:33:44.475168     1 ratelimit_manager.cpp:471] Loaded 0 rate limit rules.
I20221107 07:33:44.475193     1 ratelimit_manager.cpp:472] Loaded 0 rate limit bans.
I20221107 07:33:44.475261     1 typesense_server_utils.cpp:496] Starting API service...
I20221107 07:33:44.475533     1 http_server.cpp:178] Typesense has started listening on port 8108
I20221107 07:33:44.478169   198 typesense_server_utils.cpp:252] Since no --nodes argument is provided, starting a single node Typesense cluster.
I20221107 07:33:44.479007   199 batched_indexer.cpp:122] Starting batch indexer with 64 threads.
I20221107 07:33:44.480839   199 batched_indexer.cpp:128] BatchedIndexer skip_index: -9999
I20221107 07:33:44.575337   198 server.cpp:1045] Server[braft::RaftStatImpl+braft::FileServiceImpl+braft::RaftServiceImpl+braft::CliServiceImpl] is serving on port=8107.
I20221107 07:33:44.576820   198 server.cpp:1048] Check out http://98d69f35b38a:8107 in web browser.
I20221107 07:33:44.581704   198 raft_server.cpp:65] Nodes configuration: 172.18.0.2:8107:8108
I20221107 07:33:44.591282   198 log.cpp:674] Use crc32c as the checksum type of appending entries
I20221107 07:33:44.593061   198 log.cpp:1158] log load_meta /data/state/log/log_meta first_log_index: 45565 time: 1730
I20221107 07:33:44.595741   198 log.cpp:1098] load open segment, path: /data/state/log first_index: 45543
I20221107 07:33:44.607445   198 raft_meta.cpp:521] Loaded single stable meta, path /data/state/meta term 533 votedfor 172.18.0.3:8107:8108 time: 594
I20221107 07:33:44.608325   272 raft_server.cpp:506] on_snapshot_load
I20221107 07:33:44.622056   272 store.h:291] rm /data/db success
I20221107 07:33:44.628752   272 store.h:301] copy snapshot /data/state/snapshot/snapshot_00000000000000045565/db_snapshot to /data/db success
I20221107 07:33:44.633664   272 store.h:63] Initializing DB by opening state dir: /data/db
I20221107 07:33:44.699005   272 store.h:315] DB open success!
I20221107 07:33:44.699759   272 raft_server.cpp:485] Loading collections from disk...
I20221107 07:33:44.702265   272 collection_manager.cpp:159] CollectionManager::load()
I20221107 07:33:44.702517   272 auth_manager.cpp:32] Indexing 0 API key(s) found on disk.
I20221107 07:33:44.702626   272 collection_manager.cpp:179] Loading upto 32 collections in parallel, 1000 documents at a time.
I20221107 07:33:44.702813   272 collection_manager.cpp:186] Found 9 collection(s) on disk.
I20221107 07:33:44.707177   289 collection_manager.cpp:109] Found collection brands with 4 memory shards.
I20221107 07:33:44.719859   290 collection_manager.cpp:109] Found collection categories with 4 memory shards.
I20221107 07:33:44.725354   291 collection_manager.cpp:109] Found collection ingredient_groups with 4 memory shards.
I20221107 07:33:44.725630   295 collection_manager.cpp:109] Found collection ingredients_v2 with 4 memory shards.
I20221107 07:33:44.725713   296 collection_manager.cpp:109] Found collection marketing_claims with 4 memory shards.
I20221107 07:33:44.725791   298 collection_manager.cpp:109] Found collection posts with 4 memory shards.
I20221107 07:33:44.725931   299 collection_manager.cpp:109] Found collection products_nested with 4 memory shards.
I20221107 07:33:44.726181   300 collection_manager.cpp:109] Found collection public_products with 4 memory shards.
I20221107 07:33:44.726282   293 collection_manager.cpp:109] Found collection ingredients_nested with 4 memory shards.
I20221107 07:33:44.735705   300 collection_manager.cpp:1159] Loading collection public_products
I20221107 07:33:44.747054   289 collection_manager.cpp:1159] Loading collection brands
I20221107 07:33:44.747478   298 collection_manager.cpp:1159] Loading collection posts
I20221107 07:33:44.747330   290 collection_manager.cpp:1159] Loading collection categories
I20221107 07:33:44.748621   293 collection_manager.cpp:1159] Loading collection ingredients_nested
I20221107 07:33:44.748931   299 collection_manager.cpp:1159] Loading collection products_nested
I20221107 07:33:44.747923   295 collection_manager.cpp:1159] Loading collection ingredients_v2
I20221107 07:33:44.749214   291 collection_manager.cpp:1159] Loading collection ingredient_groups
I20221107 07:33:44.751507   296 collection_manager.cpp:1159] Loading collection marketing_claims
I20221107 07:33:44.752346   290 collection_manager.cpp:1257] Indexed 13/13 documents into collection categories
I20221107 07:33:44.752576   290 collection_manager.cpp:226] Loaded 1 collection(s) so far
I20221107 07:33:44.754590   296 collection_manager.cpp:1257] Indexed 8/8 documents into collection marketing_claims
I20221107 07:33:44.754616   296 collection_manager.cpp:226] Loaded 2 collection(s) so far
I20221107 07:33:44.866389   298 collection_manager.cpp:1257] Indexed 1116/1116 documents into collection posts
I20221107 07:33:44.866501   298 collection_manager.cpp:226] Loaded 3 collection(s) so far
I20221107 07:33:44.874645   291 collection_manager.cpp:1257] Indexed 54/54 documents into collection ingredient_groups
I20221107 07:33:44.875078   291 collection_manager.cpp:226] Loaded 4 collection(s) so far
I20221107 07:33:44.883004   289 collection_manager.cpp:1257] Indexed 1702/1702 documents into collection brands
I20221107 07:33:44.883335   289 collection_manager.cpp:226] Loaded 5 collection(s) so far
I20221107 07:33:45.487253   295 collection_manager.cpp:1257] Indexed 6815/6815 documents into collection ingredients_v2
I20221107 07:33:45.487489   295 collection_manager.cpp:226] Loaded 6 collection(s) so far
I20221107 07:33:45.657619   293 collection_manager.cpp:1257] Indexed 8975/8975 documents into collection ingredients_nested
I20221107 07:33:45.657658   293 collection_manager.cpp:226] Loaded 7 collection(s) so far
I20221107 07:34:04.085721   299 collection_manager.cpp:1257] Indexed 53185/53185 documents into collection products_nested
I20221107 07:34:04.085770   299 collection_manager.cpp:226] Loaded 8 collection(s) so far
I20221107 07:34:16.284874   300 collection_manager.cpp:1257] Indexed 44286/44286 documents into collection public_products
I20221107 07:34:16.285280   300 collection_manager.cpp:226] Loaded 9 collection(s) so far
I20221107 07:34:16.285398   272 collection_manager.cpp:263] Loaded 9 collection(s).
I20221107 07:34:16.286433   272 collection_manager.cpp:267] Initializing batched indexer from snapshot state...
I20221107 07:34:16.286567   272 batched_indexer.cpp:425] Restored 0 in-flight requests from snapshot.
I20221107 07:34:16.286607   272 raft_server.cpp:492] Finished loading collections from disk.
I20221107 07:34:16.286701   272 raft_server.h:299] Configuration of this group is 172.18.0.3:8107:8108
I20221107 07:34:16.286890   272 snapshot_executor.cpp:264] node default_group:172.18.0.2:8107:8108 snapshot_load_done, last_included_index: 45565 last_included_term: 533 peers: "172.18.0.3:8107:8108"
I20221107 07:34:16.287086   198 raft_meta.cpp:521] Loaded single stable meta, path /data/state/meta term 533 votedfor 172.18.0.3:8107:8108 time: 26
I20221107 07:34:16.287114   198 node.cpp:576] node default_group:172.18.0.2:8107:8108 init, term: 533 last_log_id: (index=45566,term=533) conf: 172.18.0.3:8107:8108 old_conf: 
I20221107 07:34:16.287145   198 raft_server.cpp:129] Node last_index: 45566
I20221107 07:34:16.287153   198 typesense_server_utils.cpp:303] Typesense peering service is running on 172.18.0.2:8107
I20221107 07:34:16.287158   198 typesense_server_utils.cpp:304] Snapshot interval configured as: 3600s
I20221107 07:34:16.287160   198 typesense_server_utils.cpp:305] Snapshot max byte count configured as: 4194304
W20221107 07:34:16.287175   198 controller.cpp:1454] SIGINT was installed with 1
I20221107 07:34:16.287185   198 raft_server.cpp:540] Term: 533, last_index index: 45566, committed_index: 0, known_applied_index: 45565, applying_index: 0, queued_writes: 0, pending_queue_size: 0, local_sequence: 10573662
W20221107 07:34:16.287191   198 raft_server.cpp:564] Single-node with no leader. Resetting peers.
W20221107 07:34:16.287195   198 node.cpp:894] node default_group:172.18.0.2:8107:8108 set_peer from 172.18.0.3:8107:8108 to 172.18.0.2:8107:8108
I20221107 07:34:16.307314   198 raft_meta.cpp:546] Saved single stable meta, path /data/state/meta term 534 votedfor 0.0.0.0:0:0 time: 20104
I20221107 07:34:21.927356   272 node.cpp:1484] node default_group:172.18.0.2:8107:8108 term 534 start pre_vote
I20221107 07:34:21.927426   272 node.cpp:1549] node default_group:172.18.0.2:8107:8108 term 534 start vote and grant vote self
I20221107 07:34:21.931159   272 raft_meta.cpp:546] Saved single stable meta, path /data/state/meta term 535 votedfor 172.18.0.2:8107:8108 time: 3596
I20221107 07:34:21.931197   272 node.cpp:1783] node default_group:172.18.0.2:8107:8108 term 535 become leader of group 172.18.0.2:8107:8108 
I20221107 07:34:21.933187   272 raft_server.h:299] Configuration of this group is 172.18.0.2:8107:8108
I20221107 07:34:21.933223   272 node.cpp:3142] node default_group:172.18.0.2:8107:8108 reset ConfigurationCtx, new_peers: 172.18.0.2:8107:8108, old_peers: 172.18.0.2:8107:8108
I20221107 07:34:21.933235   272 raft_server.h:282] Node becomes leader, term: 535

```

**Comment by kishorenc** [comment]
@reinoldus Thanks for spotting this: there's indeed a regression in a recent RC build -- I'm looking into it.

**Comment by kishorenc** [comment]
@reinoldus Can you please try against this Docker build: `typesense/typesense:0.24.0.rcn30`?

Btw, this issue only happens if the server is restarted after it has run for a while and the hourly snapshot has happened. If you index and immediately restart the server, it will not exhibit the problem you faced. You can trigger a snapshot manually and change snapshot interval via the CLI flag to a lower amount to verify the fix.

**Comment by reinoldus** [comment]
@kishorenc 

I can confirm that it's fixed! I triggered the snapshot, restarted and the facets are still there. Thank you!

**Comment by jasonbosco** [comment]
### ⚠️ Heads-up 

If you're using the new `highlight` field in the search API response in the `0.24.0.rcnX` series, we had to make a change to this field's structure in `0.24.0.rcn32`, to better accommodate different forms of nesting.

In `0.24.0.rcn1` - `0.24.0.rcn31`, the `highlight` field in the response used to look like this:

```json
{
  "hits": [
    {
      "highlight": {
        "full": {
          "categories": [
            "Cell <mark>Phone</mark>s",
            "Cell <mark>Phone</mark> Accessories",
            "SIM Cards"
          ],
          "description": "Use your compatible cell <mark>phone</mark> on the TRACFONE network with this TRACFONE TFMATKTP4C kit, which contains GSM and CDMA dual (2FF/3FF) and nano (4FF) SIM cards, activation instructions and a CDMA network access code to suit your needs.",
          "name": "TRACFONE - Bring Your Own <mark>Phone</mark> Mini Activation Kit"
        },
        "meta": {
          "categories": {
            "matched_indices": [
              0,
              1
            ],
            "matched_tokens": [
              "Phone"
            ]
          },
          "description": {
            "matched_tokens": [
              "phone"
            ]
          },
          "name": {
            "matched_tokens": [
              "Phone"
            ]
          }
        },
        "snippet": {
          "categories": [
            "Cell <mark>Phone</mark>s",
            "Cell <mark>Phone</mark> Accessories",
            "SIM Cards"
          ],
          "description": "Use your compatible cell <mark>phone</mark> on the TRACFONE network",
          "name": "TRACFONE - Bring Your Own <mark>Phone</mark> Mini Activation Kit"
        }
      }
    }
  ]
}
```

As of `v0.24.0.rcn32`, the response format now looks like this:

```json
{
  "hits": [
    {
      "highlight": {
        "categories": [
          {
            "matched_tokens": [
              "Phone"
            ],
            "snippet": "Cell <mark>Phone</mark>s",
            "value": "Cell <mark>Phone</mark>s"
          },
          {
            "matched_tokens": [
              "Phone"
            ],
            "snippet": "Cell <mark>Phone</mark> Accessories",
            "value": "Cell <mark>Phone</mark> Accessories"
          },
          {
            "matched_tokens": [],
            "snippet": "SIM Cards",
            "value": "SIM Cards"
          }
        ],
        "description": {
          "matched_tokens": [
            "phone"
          ],
          "snippet": "Use your compatible cell <mark>phone</mark>",
          "value": "Use your compatible cell <mark>phone</mark> on the TRACFONE network with this TRACFONE TFMATKTP4C kit, which contains GSM and CDMA dual (2FF/3FF) and nano (4FF) SIM cards, activation instructions and a CDMA network access code to suit your needs."
        },
        "name": {
          "matched_tokens": [
            "Phone"
          ],
          "snippet": "TRACFONE - Bring Your Own <mark>Phone</mark> Mini Activation Kit",
          "value": "TRACFONE - Bring Your Own <mark>Phone</mark> Mini Activation Kit"
        }
      }
    }
  ]
}
```

The older format required us to try and "flatten" arbitrarily nested objects / arrays, which started to get challenging to parse. This new format, just follows the same nested structure as your document, and replaces the "leaf" nodes in the nested structure (bools, strings, numeric values), with this structure:

```json
{
  "matched_tokens": ["word1", "word2"],
  "snippet": "field with just highlighted snippets",
  "value": "full field, with highlights"
}
```

**Comment by master-eugene** [comment]
@kishorenc when we may anticipate a release of 24 version with supporting nested fields?

**Comment by kishorenc** [comment]
@master-eugene We're already on code freeze now (barring some minor edge cases / bugs). So hopefully by end of the year.

**Comment by britisharmy** [comment]
I have the following php script that creates a collection that supports nesting 

```
<?php
require_once('vendor/autoload.php');
use Symfony\Component\HttpClient\HttplugClient;
use Typesense\Client;

ini_set('display_errors', '1');
ini_set('display_startup_errors', '1');
error_reporting(E_ALL);


try {
$client = new Client(
        [
            'api_key' => 'api_key',
            'nodes' => [
                [
                    'host' => 'example.com',
                    'port' => '8108',
                    'protocol' => 'https',
                ],
            ],
            'client' => new HttplugClient(),
        ]
    );

    $schema = [
    'name'      => 'les_nested',
	'enable_nested_fields'=> true,
    'fields'    => [
    [
      'name'  => 'id',
      'type'  => 'string'
    ],   	
	[
      'name'  => 'category',
      'type'  => 'string',
	  'facet' => true
    ],	
	[
      'name'  => 'boosting_score',
      'type'  => 'int32',
	  'facet' => true
    ],
	[
      'name'  => 'json_object.*',
      'type'  => 'auto',
	  'facet' => true
    ]
  ],
  'default_sorting_field' => 'boosting_score'
];

//$client->collections->create($schema);
$number = rand(2,70);
$uniq_id = uniqid();
$category = "vehicles";
$boosting_score = (int)$number;
$json = '{"files":[{"name":"IMG_20221205_164654_4.jpg","size":1889968,"url":"https://example.com/file_uploads/files/IMG_20221205_164654_4.jpg","thumbnailUrl":"https://example.com/file_uploads/files/thumbnail/IMG_20221205_164654_4.jpg","deleteUrl":"https://example.com/file_uploads/index.php?file=IMG_20221205_164654_4.jpg","deleteType":"DELETE"}],"make":"Toyota","title":"Car","mileage":45555,"year_of_manufacture":"2009","fuel_type":"CNG","transmission":"Automatic","engine_size":1200,"drive_train":"All Wheel","registration_status":"Registered","condition":"Brand New   ","color":"Beige","description":"<p>Good condition</p>","price":50000}';

$document = [
		  'id'  => "$uniq_id",
		  'category'  => "$category",
		  'boosting_score' => "$boosting_score",
		  'json_object.*'  => "$json"
];

$client->collections['les_nested']->documents->create($document);  

 		
} catch (Exception $e) {
    echo $e->getMessage();
}


?>
```

and this is my script that uses typesense without npm


```
<script>
    const typesenseInstantsearchAdapter = new TypesenseInstantSearchAdapter({
        server: {
            apiKey: 'api', // Be sure to use an API key that only allows searches, in production
            nodes: [
                {
                    host: 'keke.co.ke',
                    port: '8108',
                    protocol: 'https',
                },
            ],
        },
 
        additionalSearchParameters: {
            queryBy: 'category',
        },
    });
    const searchClient = typesenseInstantsearchAdapter.searchClient;

    const search = instantsearch({
        searchClient,
        indexName: 'les_nested',
    });

let no = JSON.parse(json_object);

    search.addWidgets([
        instantsearch.widgets.searchBox({
            container: '#searchbox',
        }),
        instantsearch.widgets.configure({
            hitsPerPage: 8,
        }),
        instantsearch.widgets.hits({
            container: '#hits',
            templates: {
                item(item) {
                    return `
                        <div>
                          <img src="${item.json_object.make}" alt="${item.json_object.model}" height="100" />
                          <div class="hit-name">
                            ${item._highlightResult.json_object.title.value}
                          </div>
                          <div class="hit-publication-year">${item.no.year_of_manufacture}</div>
                         
                        </div>
                      `;
                },
            },
        }),
        instantsearch.widgets.pagination({
            container: '#pagination',
        }),
    ]);

    search.start();
</script>
```

The object is stored as json when i peek through my collection. Is this the type of object typesense would support?

**Comment by jasonbosco** [comment]
@britisharmy

```
'json_object.*'  => "$json"
```

This converts your object to a string, so it looks like a string to Typesense. You'd need to JSON parse your object and send that object into Typesense. So for eg:

```
'json_object' => json-decode($json)
```

**Comment by britisharmy** [comment]
I did that and i am still not able to facet by a nested property.

`Uncaught (in promise) TypeError: Cannot read properties of undefined (reading 'year_of_manufacture')` from this line

`<div class="hit-publication-year">${item.json_object.year_of_manufacture}</div>`

How can i find out the version of typesense that i am running? Is there a terminal command?

I downloaded and installed this version https://dl.typesense.org/releases/0.24.0.rcn37/typesense-server-0.24.0.rcn37-amd64.deb

**Comment by jasonbosco** [comment]
Could you do a `GET /collections/les_nested` (via curl) and also `GET /collections/les_nested/documents/<documentID>` (via curl) and post the output of both? 

`GET /debug` will show you the version of Typesense running.

**Comment by britisharmy** [comment]
GET request


```
{"boosting_score":53,"category":"vehicles","id":"6393932c6554b","json_object.*":{"color":"Beige","condition":"Brand New   ","description":"<p>Good condition</p>","drive_train":"All Wheel","engine_size":1200,"files":[{"deleteType":"DELETE","deleteUrl":"https://example.com/file_uploads/index.php?file=IMG_20221205_164654_4.jpg","name":"IMG_20221205_164654_4.jpg","size":1889968,"thumbnailUrl":"https://example.com/file_uploads/files/thumbnail/IMG_20221205_164654_4.jpg","url":"https://example.com/file_uploads/files/IMG_20221205_164654_4.jpg"}],"fuel_type":"CNG","make":"Toyota","mileage":45555,"price":50000,"registration_status":"Registered","title":"Car","transmission":"Automatic","year_of_manufacture":"2009"}}

{"boosting_score":54,"category":"vehicles","id":"6393933286e29","json_object.*":{"color":"Beige","condition":"Brand New   ","description":"<p>Good condition</p>","drive_train":"All Wheel","engine_size":1200,"files":[{"deleteType":"DELETE","deleteUrl":"https://example.com/file_uploads/index.php?file=IMG_20221205_164654_4.jpg","name":"IMG_20221205_164654_4.jpg","size":1889968,"thumbnailUrl":"https://example.com/file_uploads/files/thumbnail/IMG_20221205_164654_4.jpg","url":"https://example.com/file_uploads/files/IMG_20221205_164654_4.jpg"}],"fuel_type":"CNG","make":"Toyota","mileage":45555,"price":50000,"registration_status":"Registered","title":"Car","transmission":"Automatic","year_of_manufacture":"2009"}}


{"boosting_score":17,"category":"vehicles","id":"63939335515b5","json_object.*":{"color":"Beige","condition":"Brand New   ","description":"<p>Good condition</p>","drive_train":"All Wheel","engine_size":1200,"files":[{"deleteType":"DELETE","deleteUrl":"https://example.com/file_uploads/index.php?file=IMG_20221205_164654_4.jpg","name":"IMG_20221205_164654_4.jpg","size":1889968,"thumbnailUrl":"https://example.com/file_uploads/files/thumbnail/IMG_20221205_164654_4.jpg","url":"https://example.com/file_uploads/files/IMG_20221205_164654_4.jpg"}],"fuel_type":"CNG","make":"Toyota","mileage":45555,"price":50000,"registration_status":"Registered","title":"Car","transmission":"Automatic","year_of_manufacture":"2009"}}
```

from 

```
$colls = $client->collections['decoded_object']->documents->export();
print_r($colls);
```

specific document with id

```
Array
(
    [boosting_score] => 53
    [category] => vehicles
    [id] => 6393932c6554b
    [json_object.*] => Array
        (
            [color] => Beige
            [condition] => Brand New   
            [description] => <p>Good condition</p>
            [drive_train] => All Wheel
            [engine_size] => 1200
            [files] => Array
                (
                    [0] => Array
                        (
                            [deleteType] => DELETE
                            [deleteUrl] => https://example.com/file_uploads/index.php?file=IMG_20221205_164654_4.jpg
                            [name] => IMG_20221205_164654_4.jpg
                            [size] => 1889968
                            [thumbnailUrl] => https://example.com/file_uploads/files/thumbnail/IMG_20221205_164654_4.jpg
                            [url] => https://example.com/file_uploads/files/IMG_20221205_164654_4.jpg
                        )

                )

            [fuel_type] => CNG
            [make] => Toyota
            [mileage] => 45555
            [price] => 50000
            [registration_status] => Registered
            [title] => Car
            [transmission] => Automatic
            [year_of_manufacture] => 2009
        )

)

```

I used php

```
$r = $client->collections['decoded_object']->documents['6393932c6554b']->retrieve();

print_r($r);

```

**Comment by jasonbosco** [comment]
When indexing the document you want to use the field name as `json_object` and not `json_object.*` in the document.

**Comment by britisharmy** [comment]
Many thanks. I have it now working.

**Comment by nogamenofun98** [comment]
I tried out the [0.24.0.rcn43](https://hub.docker.com/layers/typesense/typesense/0.24.0.rcn43/images/sha256-6804675edd381a65fe462131797e78d91631c06310311032223f6d91f32ba734?context=explore) and the nested field is working fine, but when I trying to put the nested field in "facet_by" like "specification", it doesnt show all the nested field under "specification" field. Right now also cannot put as "specification.*".

The sample data in "specification" field:
{
"resolution":"1920x1080",
"inch":"10inch",
}

**Comment by luke3butler** [comment]
Similar to @nogamenofun98, I'd like to add facets to fields that are inside an array of objects.

Is there an undocumented path syntax for this @jasonbosco ? 

I'm fine with defining all my fields if I have to.

**Comment by kishorenc** [comment]
@luke3butler 

You can use "." to refer to nested fields. For example, given a document like this:

```
{
     "person": {
         "name": {
              "first": "Jack"
         }
     }
}
```

You can enable facet for the internal name field by referring to the field as `person.name.first`.

**Comment by luke3butler** [comment]
@kishorenc My question is about objects inside arrays. How do those fields get referenced?

**Comment by kishorenc** [comment]
You can use the same dot notation for referencing objects inside an array as well. If you can post a specific example, I can also elaborate further.

**Comment by luke3butler** [comment]
@kishorenc 

Example object

```
{
  id: "uuid",
  name: "example",
  children: [
    {
      id: "uuid2",
      name: "example2",
      type: "type2", // <-- I want thhis to be `facet: true`
    },
    {
      id: "uuid3",
      name: "example3",
      type: "type3", // <-- and this
    }
    // And so on...
  ]
}
```
Thanks!

**Comment by alexander-schranz** [comment]
I'm currently working on a [search abstraction for PHP](https://github.com/schranz-search/schranz-search/pull/75) and want to add support for Typesense. I currently running into an unexpected error. Multiple nested objects:

My index looks like this:

<details>
<summary>Index Config JSON</summary>

```json
{
    "name": "test_complex",
    "enable_nested_fields": true,
    "fields": [
        {
            "name": "uuid",
            "type": "string"
        },
        {
            "name": "title",
            "type": "string",
            "optional": true,
            "index": true
        },
        {
            "name": "header",
            "type": "object",
            "optional": true
        },
        {
            "name": "header.image",
            "type": "object",
            "optional": true
        },
        {
            "name": "header.image.media",
            "type": "int64",
            "optional": true,
            "index": false
        },
        {
            "name": "header.video",
            "type": "object",
            "optional": true
        },
        {
            "name": "header.video.media",
            "type": "string",
            "optional": true,
            "index": false
        },
        {
            "name": "article",
            "type": "string",
            "optional": true,
            "index": true
        },
        {
            "name": "blocks",
            "type": "object",
            "optional": true
        },
        {
            "name": "blocks.text",
            "type": "object[]",
            "optional": true
        },
        {
            "name": "blocks.text._originalIndex",
            "type": "int64",
            "optional": false
        },
        {
            "name": "blocks.text.title",
            "type": "string",
            "optional": true,
            "index": true
        },
        {
            "name": "blocks.text.description",
            "type": "string",
            "optional": true,
            "index": true
        },
        {
            "name": "blocks.text.media",
            "type": "int64[]",
            "optional": true,
            "index": false
        },
        {
            "name": "blocks.embed",
            "type": "object[]",
            "optional": true
        },
        {
            "name": "blocks.embed._originalIndex",
            "type": "int64",
            "optional": false
        },
        {
            "name": "blocks.embed.title",
            "type": "string",
            "optional": true,
            "index": true
        },
        {
            "name": "blocks.embed.media",
            "type": "string",
            "optional": true,
            "index": false
        },
        {
            "name": "footer",
            "type": "object",
            "optional": true
        },
        {
            "name": "footer.title",
            "type": "string",
            "optional": true,
            "index": true
        },
        {
            "name": "created",
            "type": "int64",
            "optional": true,
            "index": true
        },
        {
            "name": "commentsCount",
            "type": "int64",
            "optional": true,
            "index": false
        },
        {
            "name": "rating",
            "type": "float",
            "optional": true,
            "index": false
        },
        {
            "name": "comments",
            "type": "object[]",
            "optional": true
        },
        {
            "name": "comments._originalIndex",
            "type": "int64",
            "optional": false
        },
        {
            "name": "comments.email",
            "type": "string",
            "optional": true,
            "index": false
        },
        {
            "name": "comments.text",
            "type": "string",
            "optional": true,
            "index": true
        },
        {
            "name": "tags",
            "type": "string[]",
            "optional": true,
            "index": true
        },
        {
            "name": "categoryIds",
            "type": "int64[]",
            "optional": true,
            "index": false
        }
    ]
}
```

</details>

And the document I want to index like this:

<details>
<summary>Document JSON</summary>

```json
{
    "uuid": "23b30f01-d8fd-4dca-b36a-4710e360a965",
    "title": "New Blog",
    "header": {
        "image": {
            "media": 1
        }
    },
    "article": "<article><h2>New Subtitle<\/h2><p>A html field with some content<\/p><\/article>",
    "blocks": {
        "text": [
            {
                "_originalIndex": 0,
                "title": "Titel",
                "description": "<p>Description<\/p>",
                "media": [
                    3,
                    4
                ]
            },
            {
                "_originalIndex": 1,
                "title": "Titel 2",
                "description": "<p>Description 2<\/p>"
            }
        ],
        "embed": [
            {
                "_originalIndex": 2,
                "title": "Video",
                "media": "https:\/\/www.youtube.com\/watch?v=iYM2zFP3Zn0"
            }
        ]
    },
    "footer": {
        "title": "New Footer"
    },
    "created": 1643022000,
    "commentsCount": 2,
    "rating": 3.5,
    "comments": [
        {
            "email": "admin.nonesearchablefield@localhost",
            "text": "Awesome blog!"
        },
        {
            "email": "example.nonesearchablefield@localhost",
            "text": "Like this blog!"
        }
    ],
    "tags": [
        "Tech",
        "UI"
    ],
    "categoryIds": [
        1,
        2
    ],
    "id": "23b30f01-d8fd-4dca-b36a-4710e360a965"
}
```

</details>

But it ends with:

```Bash
Schranz\Search\SEAL\Adapter\Typesense\Tests\TypesenseAdapterTest::testDocument
Typesense\Exceptions\RequestMalformed: Field `blocks.text.title` must be a string.

/tmp/schranz-search/packages/seal-typesense-adapter/vendor/typesense/typesense-php/src/ApiCall.php:344
/tmp/schranz-search/packages/seal-typesense-adapter/vendor/typesense/typesense-php/src/ApiCall.php:237
/tmp/schranz-search/packages/seal-typesense-adapter/vendor/typesense/typesense-php/src/ApiCall.php:125
/tmp/schranz-search/packages/seal-typesense-adapter/vendor/typesense/typesense-php/src/Documents.php:88
/tmp/schranz-search/packages/seal-typesense-adapter/TypesenseConnection.php:38
/tmp/Projects/schranz-search/packages/seal/Engine.php:31
/tmp/Projects/schranz-search/packages/seal/Testing/AbstractAdapterTestCase.php:87
```

**Comment by kishorenc** [comment]
@alexander-schranz 

The `blocks.embed.title` must be defined as an array of string (`string[]`) in the schema because `embed` is an array of objects:

```
"blocks": {
        ...
        "embed": [
            {
                "_originalIndex": 2,
                "title": "Video",
                "media": "https:\/\/www.youtube.com\/watch?v=iYM2zFP3Zn0"
            }
        ]
    },
```

**Comment by alexander-schranz** [comment]
@kishorenc Really thank you for the fast response here!

I would expected that Typesense handle that things internal to make things easier from a developer perspective. A quickfix could be to validate that in the schema / index creation so atleast there already you get a better error message what is wrong about the schema:

I got another unexpected error for the schema after required changes.

The schema looks like this now:

<details>
<summary>Schema</summary>

```json

{
    "name": "test_complex",
    "enable_nested_fields": true,
    "fields": [
        {
            "name": "uuid",
            "type": "string",
            "facet": true
        },
        {
            "name": "title",
            "type": "string",
            "optional": true,
            "index": true
        },
        {
            "name": "header",
            "type": "object",
            "optional": true
        },
        {
            "name": "header.image",
            "type": "object",
            "optional": true
        },
        {
            "name": "header.image.media",
            "type": "int64",
            "optional": true,
            "index": false
        },
        {
            "name": "header.video",
            "type": "object",
            "optional": true
        },
        {
            "name": "header.video.media",
            "type": "string",
            "optional": true,
            "index": false
        },
        {
            "name": "article",
            "type": "string",
            "optional": true,
            "index": true
        },
        {
            "name": "blocks",
            "type": "object",
            "optional": true
        },
        {
            "name": "blocks.text",
            "type": "object[]",
            "optional": true
        },
        {
            "name": "blocks.text._originalIndex",
            "type": "int64[]",
            "optional": false
        },
        {
            "name": "blocks.text.title",
            "type": "string[]",
            "optional": true,
            "index": true
        },
        {
            "name": "blocks.text.description",
            "type": "string[]",
            "optional": true,
            "index": true
        },
        {
            "name": "blocks.text.media",
            "type": "int64[]",
            "optional": true,
            "index": false
        },
        {
            "name": "blocks.embed",
            "type": "object[]",
            "optional": true
        },
        {
            "name": "blocks.embed._originalIndex",
            "type": "int64[]",
            "optional": false
        },
        {
            "name": "blocks.embed.title",
            "type": "string[]",
            "optional": true,
            "index": true
        },
        {
            "name": "blocks.embed.media",
            "type": "string[]",
            "optional": true,
            "index": false
        },
        {
            "name": "footer",
            "type": "object",
            "optional": true
        },
        {
            "name": "footer.title",
            "type": "string",
            "optional": true,
            "index": true
        },
        {
            "name": "created",
            "type": "int64",
            "optional": true,
            "index": true
        },
        {
            "name": "commentsCount",
            "type": "int64",
            "optional": true,
            "index": false
        },
        {
            "name": "rating",
            "type": "float",
            "optional": true,
            "index": false
        },
        {
            "name": "comments",
            "type": "object[]",
            "optional": true
        },
        {
            "name": "comments.email",
            "type": "string[]",
            "optional": true,
            "index": false
        },
        {
            "name": "comments.text",
            "type": "string[]",
            "optional": true,
            "index": true
        },
        {
            "name": "tags",
            "type": "string[]",
            "optional": true,
            "index": true
        },
        {
            "name": "categoryIds",
            "type": "int64[]",
            "optional": true,
            "index": false
        }
    ]
}
```

</details>

But when inserting the document without blocks:

<details>
<summary>Document.json</summary>

```json
{
    "uuid": "79848403-c1a1-4420-bcc2-06ed537e0d4d",
    "title": "Other Blog",
    "header": {
        "video": {
            "media": "https:\/\/www.youtube.com\/watch?v=iYM2zFP3Zn0"
        }
    },
    "article": "<article><h2>Other Subtitle<\/h2><p>A html field with some content<\/p><\/article>",
    "footer": {
        "title": "Other Footer"
    },
    "created": 1672052400,
    "commentsCount": 0,
    "rating": 2.5,
    "comments": [],
    "tags": [
        "UI",
        "UX"
    ],
    "categoryIds": [
        2,
        3
    ],
    "id": "79848403-c1a1-4420-bcc2-06ed537e0d4d"
}
```

</details>

It will error:

> 1) Schranz\Search\SEAL\Adapter\Typesense\Tests\TypesenseAdapterTest::testDocument
Typesense\Exceptions\RequestMalformed: Field `blocks.embed._originalIndex` has been declared in the schema, but is not found in the document. If currently not possible I would atleast check it in the `schema creation` like suggested for the array that already there a error message is returned for something unexpected like this.

Something I would not expect as `blocks` is optional. And _originalIndex should only be required when there is a `blocks.embed` data send.

**Comment by jasonbosco** [comment]
We just published the GA version of v0.24.0 today with nested fields support.

Feels great to finally close an issue as completed after iterating on it for all these months! 

A big thank you to everyone on this thread for your feedback and help getting this over the finish line 🙏

**Comment by Alexander-Zierhut** [comment]
@jasonbosco @kishorenc fantastic job! This feature is a great addition to Typesense. You make stuff actually happen and listen to the community.

![image](https://user-images.githubusercontent.com/26348781/215712900-05e7e26d-9755-4b08-91aa-4b6a11326915.png)

**Comment by karanhudia** [comment]
> I tried out the [0.24.0.rcn43](https://hub.docker.com/layers/typesense/typesense/0.24.0.rcn43/images/sha256-6804675edd381a65fe462131797e78d91631c06310311032223f6d91f32ba734?context=explore) and the nested field is working fine, but when I trying to put the nested field in "facet_by" like "specification", it doesnt show all the nested field under "specification" field. Right now also cannot put as "specification.*".
> 
> Is it possible to facet_by using nested field? or this nested field is just for indexing only?
> 
> The sample data in "specification" field: { "resolution":"1920x1080", "inch":"10inch", }

@nogamenofun98 Did you find out how to address this specific issue? Because I literally have the same issue, with the exact same field.

**Comment by jasonbosco** [comment]
@karanhudia You can indeed facet on nested fields. Could you share one sample document and tell me which field you want to facet on, and I can show you how to setup the schema for it.

**Comment by britisharmy** [comment]
Are nested objects be used as attributes in the refinement list widget like

```
instantsearch.widgets.refinementList({
		  container: '#refinement-list',
		  attribute: 'object.field',
}),
```

**Comment by jasonbosco** [comment]
Yup, should be possible if you upgrade to the latest version of the adapter and latest version of Typesense.

**Comment by britisharmy** [comment]
I am using this adapter and the latest version of typesense https://github.com/britisharmy/typesense-instantsearch-demo-no-npm-yarn

**Comment by jasonbosco** [comment]
Could you open a separate issue in the typesense-instantsearch-adapter with the exact error you running into?

**Comment by britisharmy** [comment]
done.

**Comment by nogamenofun98** [comment]
> @karanhudia You can indeed facet on nested fields. Could you share one sample document and tell me which field you want to facet on, and I can show you how to setup the schema for it.

@jasonbosco My previous comment is trying to achieve wildcard nested object facet, like I dont know what nested fields are inside the "specification" field, so I want to use "specification.*" OR "specification" in the "facet_by", and Typesense engine should return all the nested fields inside specification and the number of result of each facet. 

I think can refer to this #825 issue, I think that issue is the solution for my issue.

**Comment by itswadesh** [comment]
Hi, thanks for building such a wonderful tool. I am stuck with facets of nested fields, documentation lacs too. can you suggest why facets_count is 0
![image](https://user-images.githubusercontent.com/1191896/221107518-32ae8de2-09c0-4cda-8b3f-bb4d4c2a6c3a.png)

**Comment by kishorenc** [comment]
@itswadesh Can you please create a separate issue along with sample data and code snippet that showcases the issue. 

I'm going to lock this issue as it's for the discussion for the feature request.
